<a href="https://colab.research.google.com/github/swamhtetg90/DAVI-CA2/blob/Swam/data_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DAVI Data Cleaning
## Group-10
## Group Member: Devendran Yoheswaran, Kaung Myat San, Swam Htet Aung
---

## Meta Data
---
### Student Profiles

> **Note:** Data is manually entered, so the values are not standardized.

| **Field Name**                            | **Description**                                                                                                                          | **Example**                |
| ----------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- | -------------------------- |
| **STUDENT ID**                            | Student ID is made up of three attributes: `<Course code>-<Intake No>/<Index Number of student in the intake>`                           | `1101-013/001`             |
| **GENDER**                                | Gender                                                                                                                                   | `M`, `F`                   |
| **SG CITIZEN**                            | Singapore Citizen                                                                                                                        | `Y` or blank               |
| **SG PR**                                 | Singapore Permanent Resident                                                                                                             | `Y` or blank               |
| **FOREIGNER**                             | Neither SG Citizen nor SG PR (mutually exclusive with SG CITIZEN and SG PR)                                                              | `Y` or blank               |
| **COUNTRY OF OTHER NATIONALITY**          | Country of nationality (only for SG PR or foreigner)                                                                                     | `Malaysia`, `India`, etc.  |
| **DOB**                                   | Date of Birth. Format: `DD/MM/YYYY`                                                                                                      | `04/03/1978`               |
| **HIGHEST QUALIFICATION**                 | Highest qualification attained prior to this course                                                                                      | `Certificate`, `Diploma`   |
| **NAME OF QUALIFICATION AND INSTITUTION** | Institute where the highest qualification was attained                                                                                   | As provided by participant |
| **DATE ATTAINED HIGHEST QUALIFICATION**   | Date when the qualification was awarded. Format: `DD/MM/YYYY`                                                                            | `06/11/2016`               |
| **DESIGNATION**                           | Job designation                                                                                                                          | As provided by participant |
| **COMMENCEMENT DATE**                     | Course start date. Format: `DD/MM/YYYY`                                                                                                  | `06/01/2023`               |
| **COMPLETION DATE**                       | Course end date. Blank if course is ongoing. Format: `DD/MM/YYYY`                                                                        | `06/04/2024`               |
| **FULL-TIME OR PART-TIME**                | Whether the course is Full-time or Part-time                                                                                             | `Full-Time`, `Part-Time`   |
| **COURSE FUNDING**                        | Course funding type:<br>- `Individual`<br>- `Individual - SFC` (SkillsFuture Credit)<br>- `Sponsored`<br>- `Individual - waived App Fee` | `Individual - SFC`         |
| **REGISTRATION FEE**                      | Registration fee in SGD                                                                                                                  | As entered                 |
| **PAYMENT MODE**                          | Mode of payment                                                                                                                          | `NETS`, `Giro`, `PayNow`   |
| **COURSE FEE**                            | Course fee in SGD                                                                                                                        | As entered                 |

---

### Course Codes

| **S/N**         | **Description**        | **Example**                          |
| --------------- | ---------------------- | ------------------------------------ |
| **CODE**        | Course code (4 digits) | `1101`                               |
| **COURSE NAME** | Course name            | `Diploma in Business Administration` |

---

### Semester Results

| **S/N**        | **Description**                                                                                                           | **Example**    |
| -------------- | ------------------------------------------------------------------------------------------------------------------------- | -------------- |
| **STUDENT ID** | Must match the ID in the Student Profile dataset                                                                          | `1101-013/001` |
| **PERIOD**     | Semester number<br>• Certificate: 1 semester<br>• Diploma: 3 semesters<br>• Master’s: 2 semesters (some exceptions apply) | `1`, `2`, `3`  |
| **GPA**        | GPA for the semester.<br>• Max GPA: 4<br>• Pass GPA: Certificate/Diploma = 2, Master's = 2.3                              | `3.2`          |

---


## Importing Modules
---

In [1]:
import pandas as pd
import numpy as np

## Loading Data
---

#### Course_Code Data

In [2]:
course_code = pd.read_excel("https://github.com/swamhtetg90/DAVI-CA2/blob/main/DAVI%20CA2%20datasets%20and%20meta%20data/Course%20Codes.xlsx?raw=true")
course_code.head()

CODE                         COURSE NAME
0  1101  Diploma in Business Administration
1  1102       Diploma in Business Analytics
2  2101    Certificate in Digital Marketing
3  2102        Certificate in HR Management
4  2013   Certificate in Tourism Management

#### Semester Results Data

In [3]:
semester_results = pd.read_excel("https://github.com/swamhtetg90/DAVI-CA2/blob/main/DAVI%20CA2%20datasets%20and%20meta%20data/Semester%20Results.xlsx?raw=true")
semester_results.head()

STUDENT ID PERIOD  GPA
0  1101-009/001  Sem 1  3.5
1  1101-009/001  Sem 2  3.6
2  1101-009/001  Sem 3  3.7
3  1101-009/002  Sem 1  3.4
4  1101-009/002  Sem 2  3.5

#### Student Profiles Data

In [4]:
student_profiles = pd.read_excel("https://github.com/swamhtetg90/DAVI-CA2/blob/main/DAVI%20CA2%20datasets%20and%20meta%20data/Student%20Profiles.xlsx?raw=true")
student_profiles.head()

STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
0  1101-009/001      F              NaN         Y   
1  1101-009/002      F          Y   NaN       NaN   
2  1101-009/003      F              NaN         Y   
3  1101-009/004      F              NaN         Y   
4  1101-009/005      F          Y   NaN       NaN   

  COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
0                     Malaysia  13/09/1981           Certificate   
1                          NaN  26/07/1979           Certificate   
2                        India  01/02/1990                Degree   
3                  Netherlands  20/04/1976               Diploma   
4                          NaN  25/11/1983               Diploma   

               NAME OF QUALIFICATION AND INSTITUTION  \
0                                                SPM   
1                  Certificate in Office Skills, ITE   
2  Bachelor of Business Administration, Universit...   
3  Office Management Diploma, NCOI Rotterdam, The...   
4  Diploma in Business Admininstration, LCCI Leve...   

  DATE ATTAINED HIGHEST QUALIFICATION                  DESIGNATION  \
0                          2018-01-08         Admin & HR Assistant   
1                          2016-06-08              Admin Assistant   
2                          2015-08-08                            -   
3                          2018-02-08  HR Support / Office Manager   
4                          2015-06-08    Executive, Administration   

     COMMENCEMENT DATE      COMPLETION DATE FULL-TIME OR PART-TIME  \
0  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
1  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
2  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
3  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
4  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   

   COURSE FUNDING REGISTRATION FEE PAYMENT MODE  COURSE FEE  
0      Individual              107         GIRO        5136  
1  Individual-SFC              107         NETS        5136  
2      Individual              107         NETS        5136  
3      Individual              107         NETS        5136  
4       Sponsored              107         GIRO        4812

## Data Analysis
---

### Course Code

In [5]:
course_code.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CODE         7 non-null      int64 
 1   COURSE NAME  7 non-null      object
dtypes: int64(1), object(1)
memory usage: 244.0+ bytes


In [6]:
course_code.head(10)

CODE                                        COURSE NAME
0  1101                 Diploma in Business Administration
1  1102                      Diploma in Business Analytics
2  2101                   Certificate in Digital Marketing
3  2102                       Certificate in HR Management
4  2013                  Certificate in Tourism Management
5  5112  Specialist Diploma in Business Innovation and ...
6  5113          Specialist Diploma in Intelligent Systems

#### Course Code DataFrame Summary

- The **`Course Code` DataFrame** contains **no null values**.
- It has a total of **7 rows**.
- The data appears to be **cleaned and ready for use**.


### Student Profile
---

#### Student_ID

In [7]:
import pandas as pd

# Step 1: Get unique student IDs from each DataFrame
ids_profiles = set(student_profiles['STUDENT ID'].dropna().unique())
ids_results = set(semester_results['STUDENT ID'].dropna().unique())

# Step 2: Compare sets
if ids_profiles == ids_results:
    print("STUDENT ID columns match exactly in both datasets.")
else:
    print("Mismatch found between STUDENT ID columns.")

    # Extra: Show differences
    only_in_profiles = ids_profiles - ids_results
    only_in_results = ids_results - ids_profiles

    if only_in_profiles:
        print("Student IDs only in student_profiles:")
        print(only_in_profiles)

    if only_in_results:
        print("Student IDs only in semester_results:")
        print(only_in_results)


Mismatch found between STUDENT ID columns.
Student IDs only in student_profiles:
{'2101-111/002', '5113-009/003', '5113-009/006', '5113-009/005', '5113-009/007', '2101-111/005', '2101-111/007', '2101-111/006', '5113-009/004', '2101-111/003', '5113-009/002', '2101-111/001', '2101-111/008', '2101-111/004', '5113-009/001'}
Student IDs only in semester_results:
{'5112-007/001', '2101-106/001', '2101-106/004', '2101-106/003', '5112-007/002', '5112-007/006', '2101-106/005', '5112-007/004', '2101-106/002', '5112-007/003', '5112-007/005'}


In [8]:
# Get sets of Student IDs
ids_profiles = set(student_profiles['STUDENT ID'].dropna().unique())
ids_results = set(semester_results['STUDENT ID'].dropna().unique())

# Find mismatched IDs (present in semester_results but not in student_profiles)
only_in_results = ids_results - ids_profiles

# Filter and print those rows from semester_results
mismatched_rows = semester_results[semester_results['STUDENT ID'].isin(only_in_results)]

print("❌ Mismatched rows from semester_results:")
print(mismatched_rows)


❌ Mismatched rows from semester_results:
       STUDENT ID PERIOD  GPA
116  2101-106/001  Sem 1  2.4
117  2101-106/002  Sem 1  3.1
118  2101-106/003  Sem 1  3.4
119  2101-106/004  Sem 1  2.8
120  2101-106/005  Sem 1  2.3
130  5112-007/001  Sem 1  2.9
131  5112-007/001  Sem 2  3.8
132  5112-007/002  Sem 1  3.2
133  5112-007/002  Sem 2  3.3
134  5112-007/003  Sem 1  3.6
135  5112-007/003  Sem 2  3.7
136  5112-007/004  Sem 1  3.5
137  5112-007/004  Sem 2  2.6
138  5112-007/005  Sem 1  2.1
139  5112-007/005  Sem 2  2.3
140  5112-007/006  Sem 1  3.2
141  5112-007/006  Sem 2  3.1


In [9]:
student_profiles

STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
0    1101-009/001      F              NaN         Y   
1    1101-009/002      F          Y   NaN       NaN   
2    1101-009/003      F              NaN         Y   
3    1101-009/004      F              NaN         Y   
4    1101-009/005      F          Y   NaN       NaN   
..            ...    ...        ...   ...       ...   
302  5113-009/003      F          Y   NaN       NaN   
303  5113-009/004      M        NaN     Y       NaN   
304  5113-009/005      F          Y   NaN       NaN   
305  5113-009/006      F        NaN     Y       NaN   
306  5113-009/007      F          Y   NaN       NaN   

    COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
0                       Malaysia  13/09/1981           Certificate   
1                            NaN  26/07/1979           Certificate   
2                          India  01/02/1990                Degree   
3                    Netherlands  20/04/1976               Diploma   
4                            NaN  25/11/1983               Diploma   
..                           ...         ...                   ...   
302                          NaN  06/03/1995                Degree   
303                  Philippines  09/02/1995                Degree   
304                          NaN  01/10/1993                Degree   
305                    Indonesia  30/09/1996                Degree   
306                          NaN  06/11/1996                Degree   

                 NAME OF QUALIFICATION AND INSTITUTION  \
0                                                  SPM   
1                    Certificate in Office Skills, ITE   
2    Bachelor of Business Administration, Universit...   
3    Office Management Diploma, NCOI Rotterdam, The...   
4    Diploma in Business Admininstration, LCCI Leve...   
..                                                 ...   
302  Bachelor of Science in Hotel Administration (H...   
303  Bachelor of Science in Accounting and Finance ...   
304  Bachelor of Commerce (Accounting and Finance)/...   
305  Bachelor of Economics (Accounting)/\nUniversit...   
306  Degree of Computer Science, National Universit...   

    DATE ATTAINED HIGHEST QUALIFICATION                  DESIGNATION  \
0                            2018-01-08         Admin & HR Assistant   
1                            2016-06-08              Admin Assistant   
2                            2015-08-08                            -   
3                            2018-02-08  HR Support / Office Manager   
4                            2015-06-08    Executive, Administration   
..                                  ...                          ...   
302                          2019-06-08                   HR Manager   
303                          2020-08-05              Finance Manager   
304                          2019-11-05                  Head, Admin   
305                          2018-05-05                   Accountant   
306                          2021-04-20          Software Consultant   

       COMMENCEMENT DATE      COMPLETION DATE FULL-TIME OR PART-TIME  \
0    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
1    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
2    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
3    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
4    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
..                   ...                  ...                    ...   
302  2025-04-14 00:00:00                  NaN              Part-Time   
303  2025-04-14 00:00:00                  NaN              Part-Time   
304  2025-04-14 00:00:00                  NaN              Part-Time   
305  2025-04-14 00:00:00                  NaN              Part-Time   
306  2025-04-14 00:00:00                  NaN              Part-Time   

     COURSE FUNDING REGISTRATION FEE PAYMENT MODE  COURSE FEE  
0        Indiv

In [10]:
# From Ben, i dont think we should remove students which are not found in semester_results dataset as when i checked the data,
# i found out it is because they just started the course so it could be useful for finding out no of intake over time
# Look Below

import pandas as pd

# Extract all unique prefixes from semester_results
prefixes = set(semester_results['STUDENT ID'].str.extract(r'^(\d{4}-\d{3})')[0])

# Filter student_profiles to include only rows whose STUDENT ID starts with one of the prefixes
matched_profiles = student_profiles[
    student_profiles['STUDENT ID'].str.extract(r'^(\d{4}-\d{3})')[0].isin(prefixes)
].copy()

# Display result
print("✅ Matched student profiles (by prefix only):")
matched_profiles


✅ Matched student profiles (by prefix only):


STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
0    1101-009/001      F              NaN         Y   
1    1101-009/002      F          Y   NaN       NaN   
2    1101-009/003      F              NaN         Y   
3    1101-009/004      F              NaN         Y   
4    1101-009/005      F          Y   NaN       NaN   
..            ...    ...        ...   ...       ...   
295  5113-008/003      F          Y   NaN       NaN   
296  5113-008/004      F          Y   NaN       NaN   
297  5113-008/005      F          Y   NaN       NaN   
298  5113-008/006      F          Y   NaN       NaN   
299  5113-008/007      F        NaN   NaN         Y   

    COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
0                       Malaysia  13/09/1981           Certificate   
1                            NaN  26/07/1979           Certificate   
2                          India  01/02/1990                Degree   
3                    Netherlands  20/04/1976               Diploma   
4                            NaN  25/11/1983               Diploma   
..                           ...         ...                   ...   
295                          NaN  18/11/1991                Degree   
296                          NaN  29/04/1974                Degree   
297                          NaN  19/10/1981                Degree   
298                          NaN  19/03/1971                Degree   
299                        China  28/11/1998                Degree   

                 NAME OF QUALIFICATION AND INSTITUTION  \
0                                                  SPM   
1                    Certificate in Office Skills, ITE   
2    Bachelor of Business Administration, Universit...   
3    Office Management Diploma, NCOI Rotterdam, The...   
4    Diploma in Business Admininstration, LCCI Leve...   
..                                                 ...   
295  Bachelor of Business (Marketing)/\nRMIT Univer...   
296  Bachelor of Commerce in Management and Marketi...   
297  Bachelor of Arts in Human Resource Management,...   
298  Bachelor of Arts (Sociology)/\nState Universit...   
299  Bachelor of Arts with Honours/\nNational Unive...   

    DATE ATTAINED HIGHEST QUALIFICATION                   DESIGNATION  \
0                            2018-01-08          Admin & HR Assistant   
1                            2016-06-08               Admin Assistant   
2                            2015-08-08                             -   
3                            2018-02-08   HR Support / Office Manager   
4                            2015-06-08     Executive, Administration   
..                                  ...                           ...   
295                          2017-03-21            Regional Recruiter   
296                          2017-02-28        Confidential Assistant   
297                          2018-01-30  Journey Management Team Lead   
298                          1995-05-30   Academy Program Coordinator   
299                          2019-11-30          Management Associate   

       COMMENCEMENT DATE      COMPLETION DATE FULL-TIME OR PART-TIME  \
0    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
1    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
2    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
3    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
4    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
..                   ...                  ...                    ...   
295  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   
296  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   
297  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   
298  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   
299  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   

     COURSE FUNDING REGISTRATION FEE PAYMENT MODE  COURSE FEE  
0 

307 - 292 = 15 rows are mismatched.



---

### Checking students which are not found in semester_results

In [11]:
# Get sets of Student IDs
ids_profiles = set(student_profiles['STUDENT ID'].dropna().unique())
ids_results = set(semester_results['STUDENT ID'].dropna().unique())

# Find mismatched IDs (present in student_profiles but not in student_profiles)
only_in_results =  ids_profiles - ids_results

# Filter and print those rows from student_profiles
mismatched_rows = student_profiles[student_profiles['STUDENT ID'].isin(only_in_results)]

print("❌ Mismatched rows from student_profiles:")
mismatched_rows.head(20)


❌ Mismatched rows from student_profiles:


STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
108  2101-111/001      F                Y       NaN   
109  2101-111/002      F          Y   NaN       NaN   
110  2101-111/003      F          Y   NaN       NaN   
111  2101-111/004      F          Y   NaN       NaN   
112  2101-111/005      F          Y   NaN       NaN   
113  2101-111/006      F          Y   NaN       NaN   
114  2101-111/007      F          Y   NaN       NaN   
115  2101-111/008      F          Y   NaN       NaN   
300  5113-009/001      F        NaN   NaN         Y   
301  5113-009/002      F          Y   NaN       NaN   
302  5113-009/003      F          Y   NaN       NaN   
303  5113-009/004      M        NaN     Y       NaN   
304  5113-009/005      F          Y   NaN       NaN   
305  5113-009/006      F        NaN     Y       NaN   
306  5113-009/007      F          Y   NaN       NaN   

    COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
108                    Malaysian  08/07/1995                Degree   
109                          NaN  08/09/1997           Certificate   
110                               19/06/1999               Diploma   
111                               28/09/2010           Certificate   
112                          NaN  19/10/1990           Certificate   
113                               10/04/1999           Certificate   
114                               24/05/2000           Certificate   
115                               12/04/2001           Certificate   
300                        India  14/11/1985                Degree   
301                          NaN  28/01/1993                Degree   
302                          NaN  06/03/1995                Degree   
303                  Philippines  09/02/1995                Degree   
304                          NaN  01/10/1993                Degree   
305                    Indonesia  30/09/1996                Degree   
306                          NaN  06/11/1996                Degree   

                 NAME OF QUALIFICATION AND INSTITUTION  \
108  Bachelor of Science (HRD), Universiti Teknolog...   
109  Certificate in Grammar & Writing Intermediate ...   
110  Diploma in Mechatronic Engineering, Nee Ann Po...   
111        Higher Nitec in Hospitality Operations, ITE   
112                                           O' level   
113  Higher Nitec in Business Studies (Service Mana...   
114                                          O' levels   
115                                          O' levels   
300  Bachelor of Arts in Business with Logistics an...   
301  Bachelor of Business (Business Administration)...   
302  Bachelor of Science in Hotel Administration (H...   
303  Bachelor of Science in Accounting and Finance ...   
304  Bachelor of Commerce (Accounting and Finance)/...   
305  Bachelor of Economics (Accounting)/\nUniversit...   
306  Degree of Computer Science, National Universit...   

    DATE ATTAINED HIGHEST QUALIFICATION  \
108                          2020-04-04   
109                          2021-10-04   
110                          2020-12-24   
111                          2021-09-24   
112                          2010-07-24   
113                          2012-04-24   
114                          2020-04-24   
115                          2022-05-24   
300                          2019-08-04   
301                          2020-10-04   
302                          2019-06-08   
303                          2020-08-05   
304                          2019-11-05   
305                          2018-05-05   
306                          2021-04-20   

                                  DESIGNATION    COMMENCEMENT DATE  \
108                                   Manager  2025-04-24 00:00:00   
109                           Admin Assistant  2025-04-24 00:00:00   
110                                HR Manager  2025-04-24 00:00:00   
111                           Admin Executive  2025-04-24 00:00:00   
112                           Admin E

Students that are included in student_profile dataset but not in semester_results are those who haven't been graded yet.

#### Gender
---

In [12]:
# Check unique values in the 'GENDER' column
unique_genders = student_profiles['GENDER'].value_counts().reset_index()
unique_genders.columns = ['Gender', 'Count']
print("Unique values in 'GENDER' column:")
display(unique_genders)

Unique values in 'GENDER' column:


Gender  Count
0      F    265
1      M     42

From looking at the unique values, it seemed to be clean.

#### SG CITIZEN, SG PR & FOREIGNER
---

In [13]:
import pandas as pd
import numpy as np

# Load the student_profiles DataFrame
student_profiles = pd.read_excel("https://github.com/swamhtetg90/DAVI-CA2/blob/main/DAVI%20CA2%20datasets%20and%20meta%20data/Student%20Profiles.xlsx?raw=true")

# Replace empty strings with NaN in the specified columns
cols_to_check = ['SG CITIZEN', 'SG PR', 'FOREIGNER']
for col in cols_to_check:
    student_profiles[col] = student_profiles[col].replace(r'^\s*$', np.nan, regex=True)

# Check if there is only one non-null value among 'SG CITIZEN', 'SG PR', and 'FOREIGNER' for each row
student_profiles['Residential_Status_Check'] = (student_profiles[['SG CITIZEN', 'SG PR', 'FOREIGNER']].notna().sum(axis=1) == 1)

# Print rows where the condition is False (i.e., not exactly one non-null value)
mismatched_residential_status = student_profiles[student_profiles['Residential_Status_Check'] == False]

if mismatched_residential_status.empty:
    print("Each student has exactly one residential status specified.")
else:
    print("The following rows have more or less than one residential status specified:")
    display(mismatched_residential_status[['STUDENT ID', 'SG CITIZEN', 'SG PR', 'FOREIGNER', 'Residential_Status_Check']])

# Drop the temporary check column
student_profiles = student_profiles.drop(columns=['Residential_Status_Check'])

Each student has exactly one residential status specified.


Therefore, we can combine these 3 columns into 1 columnn called "Residential Status".

#### Nationality
---

In [14]:
# Count non-null unique values
unique_nationalities = student_profiles['COUNTRY OF OTHER NATIONALITY'].value_counts(dropna=False).reset_index()
unique_nationalities.columns = ['Nationality', 'Count']

# Separate null count from the rest
null_count = unique_nationalities[unique_nationalities['Nationality'].isna()]['Count'].values[0] if unique_nationalities['Nationality'].isna().any() else 0

# Drop the NaN row to show only actual nationalities in the list
non_null_nationalities = unique_nationalities[unique_nationalities['Nationality'].notna()].copy()

# Display summary
print("Unique non-null values in 'COUNTRY OF OTHER NATIONALITY' column:")
display(non_null_nationalities)

print(f"\nNumber of null (missing) values: {null_count}")
print(f"Total number of records: {len(student_profiles)}")


Unique non-null values in 'COUNTRY OF OTHER NATIONALITY' column:


Nationality  Count
1                   88
2      Malaysia     27
3         China     14
4         India     12
5   Philippines      8
6       Myanmar      3
7   Netherlands      1
8     Malaysian      1
9       Vietnam      1
10    Indonesia      1


Number of null (missing) values: 151
Total number of records: 307


Based on the meta data given, it is said that for SG citizens, it will show blank (null values) in this column, so we need to check whether number of null values match number of 'Y' in SG CITIZEN column.

In [15]:
# Count occurrences including NaN (null) values
sg_citizen_counts = student_profiles['SG CITIZEN'].value_counts(dropna=False).reset_index()

# Rename columns for clarity
sg_citizen_counts.columns = ['SG CITIZEN Value', 'Count']

# Display the counts
print("Count of each value in the 'SG CITIZEN' column (including nulls):")
display(sg_citizen_counts)


Count of each value in the 'SG CITIZEN' column (including nulls):


SG CITIZEN Value  Count
0                Y    224
1              NaN     68
2              Yes     15

Based on the results, you can see that if it is blank in "COUNTRY OF OTHER NATIONALITY", the Nationality is Singapore

#### DOB (Date of Birth)
---

In [16]:
# Check for missing values in the 'DOB' column
missing_dob = student_profiles['DOB'].isnull().sum()
print(f"Number of missing values in 'DOB': {missing_dob}")

# Attempt to convert 'DOB' to datetime, coercing errors
student_profiles['DOB_datetime'] = pd.to_datetime(student_profiles['DOB'], errors='coerce', format='%d/%m/%Y')

# Check for values that couldn't be parsed (will be NaT - Not a Time)
invalid_dob = student_profiles[student_profiles['DOB_datetime'].isna() & student_profiles['DOB'].notna()]

print("\nRows with invalid 'DOB' values:")
display(invalid_dob[['STUDENT ID', 'DOB', 'DOB_datetime']])

# Drop the temporary datetime column
student_profiles = student_profiles.drop(columns=['DOB_datetime'])

Number of missing values in 'DOB': 0

Rows with invalid 'DOB' values:


STUDENT ID          DOB DOB_datetime
105   2101-110/006  13-Feb-1984          NaT
106   2101-110/007  13-Jul-1987          NaT
107   2101-110/008  15-Jul-1994          NaT
184  2102-067A/011   16-07-1991          NaT

From looking at values that cannot be chnged to datetime, it shows that month column is in its short month abbreviation form instead of number.

So, we will have to convert it.

#### Highest Qualification
----

In [17]:
# Step 1: Create a new column with the count of commas in each row
student_profiles['Comma Count'] = student_profiles['NAME OF QUALIFICATION AND INSTITUTION'].astype(str).str.count(',')

# Step 2: Group by the number of commas and count rows in each group
comma_group_counts = student_profiles['Comma Count'].value_counts().sort_index().reset_index()
comma_group_counts.columns = ['Number of Commas', 'Number of Rows']

# Step 3: Display the result
print("Rows grouped by number of commas in 'NAME OF QUALIFICATION AND INSTITUTION':")
display(comma_group_counts)


Rows grouped by number of commas in 'NAME OF QUALIFICATION AND INSTITUTION':


Number of Commas  Number of Rows
0                 0             137
1                 1             143
2                 2              21
3                 3               6

#### Data Analysis Based on Comma Count

- **Rows with 1 comma (143 rows)**:
  - Clear structure for data splitting
  - Can use comma "," as delimiter to split data into qualification and institute
  - Will implement this splitting approach for these rows

- **Rows with 0 commas**:
  - Expected to contain only qualification name OR institute name
  - Require further examination to determine content type
  - Need additional analysis to classify these entries

- **Rows with 2 or more commas**:
  - Complex structure requiring detailed investigation
  - Need further examination to understand data format
  - May require custom parsing logic for proper data extraction

#### Rows with 0 commas

In [33]:
# Filter rows with no commas (Comma Count == 0) and reset index
no_comma_rows = student_profiles[student_profiles['Comma Count'] == 0].reset_index(drop=True)

# Display the filtered rows
print("Rows with NO commas in 'NAME OF QUALIFICATION AND INSTITUTION':")
display(no_comma_rows[['NAME OF QUALIFICATION AND INSTITUTION']])



Rows with NO commas in 'NAME OF QUALIFICATION AND INSTITUTION':


NAME OF QUALIFICATION AND INSTITUTION
0                                                  SPM
1                                             N' level
2                                             O' level
3                                             N' Level
4                                                  SPM
..                                                 ...
132  Bachelor of Arts in Business with Logistics an...
133  Bachelor of Business (Business Administration)...
134  Bachelor of Science in Accounting and Finance ...
135  Bachelor of Commerce (Accounting and Finance)/...
136  Bachelor of Economics (Accounting)/\nUniversit...

[137 rows x 1 columns]

From this table, I've observed that most of them actually contain institution names, but are connected by "/" instead of commas, so for these rows, I will again count number of "/".

In [19]:
# Count '/' in rows that have no commas
no_comma_rows['Slash Count'] = no_comma_rows['NAME OF QUALIFICATION AND INSTITUTION'].str.count('/')

# Group and count how many rows have how many slashes
slash_count_summary = no_comma_rows['Slash Count'].value_counts().sort_index().reset_index()
slash_count_summary.columns = ['Number of Slashes', 'Number of Rows']

print("Distribution of '/' in rows WITHOUT commas:")
display(slash_count_summary)


Distribution of '/' in rows WITHOUT commas:


Number of Slashes  Number of Rows
0                  0              57
1                  1              80

Rows with 0 slashes contain only qualification names, and institues are blank. What to do? for rows with 0 comma, check for slashes, if got 0 slashes-> confirm qualification, if got 1 slash, separate by slash.

In [43]:
# Count '/' in rows that have no commas
no_comma_rows['Slash Count'] = no_comma_rows['NAME OF QUALIFICATION AND INSTITUTION'].str.count('/')

# Filter rows in no_comma_rows that contain at least one slash
rows_with_slashes_in_no_comma = no_comma_rows[no_comma_rows['Slash Count'] > 0].reset_index(drop=True)

# Display the filtered rows
print("Rows with slashes in 'NAME OF QUALIFICATION AND INSTITUTION' (from no_comma_rows):")
display(rows_with_slashes_in_no_comma[['NAME OF QUALIFICATION AND INSTITUTION']])

Rows with slashes in 'NAME OF QUALIFICATION AND INSTITUTION' (from no_comma_rows):


NAME OF QUALIFICATION AND INSTITUTION
0   Degree of Bachelor of Science with Honours in ...
1   Degree of Bachelor of Science with Honours in ...
2   Honours Degree of Bachelor of Science (Managem...
3   Master of Arts (Chinese Studies)/\nNational Un...
4   Bachelor of Business (Accounting)/\nMurdoch Un...
..                                                ...
75  Bachelor of Arts in Business with Logistics an...
76  Bachelor of Business (Business Administration)...
77  Bachelor of Science in Accounting and Finance ...
78  Bachelor of Commerce (Accounting and Finance)/...
79  Bachelor of Economics (Accounting)/\nUniversit...

[80 rows x 1 columns]

#### Rows with 1 comma

In [34]:
# Filter rows with no commas (Comma Count == 1) and reset index
one_comma_rows = student_profiles[student_profiles['Comma Count'] == 1].reset_index(drop=True)

# Display the filtered rows
print("Rows with one comma in 'NAME OF QUALIFICATION AND INSTITUTION':")
display(one_comma_rows[['NAME OF QUALIFICATION AND INSTITUTION']])


Rows with one comma in 'NAME OF QUALIFICATION AND INSTITUTION':


NAME OF QUALIFICATION AND INSTITUTION
0                    Certificate in Office Skills, ITE
1                  Lower Secondary Education, Malaysia
2    Bahcelor of Management (Marketing), Universiti...
3    Diploma in Multimedia & Infocomm Technology, N...
4             Diploma in Business, Temasek Polytechnic
..                                                 ...
138                             BBA, Monash University
139                             BBA, Monash University
140  Bachelor of Arts in Human Resource Management,...
141  Bachelor of Science in Hotel Administration (H...
142  Degree of Computer Science, National Universit...

[143 rows x 1 columns]

There are some pecularity in these rows, some don't even contain institution names, just the country, some use comma as a separator between university and country. I will first rows with slashes.

In [21]:
# Count '/' in rows that have no commas
one_comma_rows['Slash Count'] = one_comma_rows['NAME OF QUALIFICATION AND INSTITUTION'].str.count('/')

# Group and count how many rows have how many slashes
slash_count_summary = one_comma_rows['Slash Count'].value_counts().sort_index().reset_index()
slash_count_summary.columns = ['Number of Slashes', 'Number of Rows']

print("Distribution of '/' in rows WITHOUT commas:")
display(slash_count_summary)

Distribution of '/' in rows WITHOUT commas:


Number of Slashes  Number of Rows
0                  0             137
1                  1               6

In [35]:


# Further filter those with exactly one slash
one_slash_in_one_comma = one_comma_rows[one_comma_rows['NAME OF QUALIFICATION AND INSTITUTION'].str.count('/') == 1].reset_index(drop=True)

# Display the result
print("Rows with EXACTLY 1 comma and EXACTLY 1 slash in 'NAME OF QUALIFICATION AND INSTITUTION':")
display(one_slash_in_one_comma[['NAME OF QUALIFICATION AND INSTITUTION']])


Rows with EXACTLY 1 comma and EXACTLY 1 slash in 'NAME OF QUALIFICATION AND INSTITUTION':


NAME OF QUALIFICATION AND INSTITUTION
0  SPM, LCCI Level 2 Certificate in Book-keeping ...
1  Advanced Diploma in Business Administration, U...
2  Bachelor of Accountancy, jointly offered by NU...
3  Professional Diploma in Leadership and People ...
4  Bachelor of Science in Hotel Administration (H...
5  Bachelor of Science in Hotel Administration (H...



#### Data Separation Process:
- **Split by both comma AND slash** to get 3 components

#### Component Classification Rules:

- **1st Component**:
  - Always confirmed as **qualification**
  - No further analysis needed

- **2nd Component**:
  - **Check 1**: Contains "university" or "academy" or "institute", if not OR contains at least one word in ALL CAPS along with "offered"
    - **If YES** → Classify as **institution**
  - **Check 2**: If Check 1 fails, check if contains city or country name
    - **If YES** → Put under **"country where qualification is gained" column**
  - **Check 3**: If both checks fail → Classify as **qualification**

- **3rd Component**:
  - **Check 1**: Contains "university" or "academy" or "institute" if not OR contains at least one word in ALL CAPS along with "offered"
    - **If YES** → Classify as **institution**
  - **Check 2**: If Check 1 fails, check if contains city or country name
    - **If YES** → Put under **"country where qualification is gained" column**
  - **Check 3**: If both checks fail → Classify as **qualification**

#### Rows with 2 commas

In [38]:
# Filter rows with no commas (Comma Count == 1) and reset index
two_comma_rows = student_profiles[student_profiles['Comma Count'] == 2].reset_index(drop=True)

# Display the filtered rows
print("Rows with one comma in 'NAME OF QUALIFICATION AND INSTITUTION':")
display(two_comma_rows[['NAME OF QUALIFICATION AND INSTITUTION']])

Rows with one comma in 'NAME OF QUALIFICATION AND INSTITUTION':


NAME OF QUALIFICATION AND INSTITUTION
0   Bachelor of Business Administration, Universit...
1   Office Management Diploma, NCOI Rotterdam, The...
2   Advanced Diploma in Tourism, Hospitality and E...
3   Certified Accounting Technicians (CATS), ACCA,...
4   Certified Accounting Technicians (CATS), ACCA,...
5                                NTC 2, N Level, WPLN
6   WSQ Higher Certificate in Human Resources, WPL...
7   Certificate in Payroll Administration, SHRI Ac...
8                                NTC 2, N Level, WPLN
9   SPM, Certificate in Business Studies (Business...
10  O' level, Foundation Certificate in Human Reso...
11  N Level, Foundation Certificate in HRM, SHRI A...
12  Nitec in Electronics, Computer Networking & Co...
13    Diploma in Human Resource Management, PSB, ACTA
14  Diploma in Electronics, Computer & Communicati...
15  Master of Business Administration (HR), Amity ...
16  Bachelor of Business Administration, St Paul C...
17  Bachelor of Business Administration, Xi'an Shi...
18  Certificate in Opticianry Practice, Republic P...
19       O level, Private Secretary Certificate, LCCI
20  Master of Business Administration, Anna Univer...

In [39]:
# Count '/' in rows that have two commas
two_comma_rows['Slash Count'] = two_comma_rows['NAME OF QUALIFICATION AND INSTITUTION'].str.count('/')

# Group and count how many rows have how many slashes
slash_count_summary_two_commas = two_comma_rows['Slash Count'].value_counts().sort_index().reset_index()
slash_count_summary_two_commas.columns = ['Number of Slashes', 'Number of Rows']

print("Distribution of '/' in rows with TWO commas:")
display(slash_count_summary_two_commas)

Distribution of '/' in rows with TWO commas:


Number of Slashes  Number of Rows
0                  0              21

Qualification Data Parsing Logic

Core Principles
1. **First part = Always qualification** (confirmed)
2. **All three columns can have multiple values** (join with commas)
3. **Categorize remaining parts** using keyword matching
4. **Group by category** and join multiple values

Updated Keyword Categories

1. Qualification Keywords
- **Degree types**: `Bachelor`, `Master`, `Diploma`, `Certificate`, `Degree`, `PhD`, `Doctorate`
- **Education levels**: `O Level`, `A Level`, `N Level`, `NTC`, `Nitec`, `WSQ`
- **Certifications**: `Certified`, `Advanced`, `Higher`, `Foundation`
- **Qualification codes**: `SPM`, `WPLN`, `CATS`, `ACTA`, `LCCI`

2. Institute Keywords
- **Full names**: `University`, `College`, `Polytechnic`, `Academy`, `Institute`, `School`, `Poly`
- **Known institutes**: `ITE`, `PSB`, `MDIS`, `ACCA`, `SHRI`, `NCOI`
- **Specific institutes**: `Singapore Polytechnic`, `Republic Poly`, `Tunku Abdul Rahman College`

3. Country Keywords
- `India`, `China`, `Philippines`, `The Netherlands`, `Netherlands`, `Singapore`

Processing Logic

Step 1: Split and Initialize
- Split by commas into parts
- First part → Qualification column
- Initialize empty lists for qualifications, institutes, countries

Step 2: Categorize Remaining Parts
For each remaining part:
- **Check country keywords first** (exact or partial match)
- **Check institute keywords** (contains any keyword)
- **Check qualification keywords** (contains any keyword)
- **Default**: If all caps and short (≤10 chars) → treat as qualification

Step 3: Handle Multiple Values
- **Qualifications**: Join all qualification parts with commas
- **Institutes**: Join all institute parts with commas
- **Countries**: Join all country parts with commas
- **Set null** if no values found for a category

Final Examples

| Original String | Qualification | Institute | Country |
|----------------|---------------|-----------|---------|
| `Bachelor of Business Administration, University of Rajasthan, India` | Bachelor of Business Administration | University of Rajasthan | India |
| `NTC 2, N Level, WPLN` | NTC 2, N Level, WPLN | null | null |
| `O level, Private Secretary Certificate, LCCI` | O level, Private Secretary Certificate, LCCI | null | null |
| `SPM, Certificate in Business Studies Level IV, Tunku Abdul Rahman College` | SPM, Certificate in Business Studies Level IV | Tunku Abdul Rahman College | null |
| `Diploma in Human Resource Management, PSB, ACTA` | Diploma in Human Resource Management, ACTA | PSB | null |
| `Certified Accounting Technicians (CATS), ACCA, SPM` | Certified Accounting Technicians (CATS), SPM | ACCA | null |

Priority Rules
1. **Country detection** has highest priority (usually last part)
2. **Institute detection** for clear institutional names
3. **Qualification detection** for educational levels and certifications
4. **Ambiguous short caps** → default to qualification
5. **Join multiple values** in same category with commas

#### Rows with 3 commas

In [40]:
# Filter rows with 3 commas (Comma Count == 3) and reset index
three_comma_rows = student_profiles[student_profiles['Comma Count'] == 3].reset_index(drop=True)

# Display the filtered rows
print("Rows with three commas in 'NAME OF QUALIFICATION AND INSTITUTION':")
display(three_comma_rows[['NAME OF QUALIFICATION AND INSTITUTION']])

Rows with three commas in 'NAME OF QUALIFICATION AND INSTITUTION':


NAME OF QUALIFICATION AND INSTITUTION
0  Diploma in Business Admininstration, LCCI Leve...
1  O' levels, Diploma in Business Management (dis...
2  Diploma in Business Admininstration, LCCI Leve...
3  Diploma in Financial Informatics, Nanyang Poly...
4  Bachelor of Arts, New Era University, Quezon C...
5  Diploma in Business Administration (HRM), PSB ...

In [41]:
# Count '/' in rows that have three commas
three_comma_rows['Slash Count'] = three_comma_rows['NAME OF QUALIFICATION AND INSTITUTION'].str.count('/')

# Group and count how many rows have how many slashes
slash_count_summary_three_commas = three_comma_rows['Slash Count'].value_counts().sort_index().reset_index()
slash_count_summary_three_commas.columns = ['Number of Slashes', 'Number of Rows']

print("Distribution of '/' in rows with THREE commas:")
display(slash_count_summary_three_commas)

Distribution of '/' in rows with THREE commas:


Number of Slashes  Number of Rows
0                  0               5
1                  2               1

Final Complete Qualification Data Parsing Logic

Core Processing Steps

Step 1: Split by Multiple Delimiters
- **Split by both commas AND slashes**: `,` and `/`
- **Clean each part**: Remove extra spaces, quotes, brackets
- **First part = Always qualification**

Step 2: Updated Keyword Categories

Qualification Keywords
- **Degree types**: `Bachelor`, `Master`, `Diploma`, `Certificate`, `Degree`, `PhD`, `Doctorate`
- **Education levels**: `O Level`, `A Level`, `N Level`, `NTC`, `Nitec`, `WSQ`
- **Certifications**: `Certified`, `Advanced`, `Higher`, `Foundation`
- **Professional codes**: `SPM`, `WPLN`, `CATS`, `ACTA`, `LCCI`, `CPA`, `CHRM`

Institute Keywords
- **Full names**: `University`, `College`, `Polytechnic`, `Academy`, `Institute`, `School`, `Poly`
- **Known institutes**: `ITE`, `PSB`, `MDIS`, `ACCA`, `SHRI`, `NCOI`
- **Specific institutes**: `Singapore Polytechnic`, `Republic Poly`, `Nanyang Polytechnic`, `New Era University`, `PSB Academy`

Country Keywords
- `India`, `China`, `Philippines`, `The Netherlands`, `Netherlands`, `Singapore`, `Quezon City` (treat as location indicator for Philippines)

City Keywords (Treat as Institutes)
- `Glasgow`, `Dublin`, `Quezon City`, `Rotterdam`

Step 3: Categorization Logic
For each part after the first:
1. **Check country keywords** (exact match or contains country name)
2. **Check institute keywords** (contains any institute keyword OR is a known city)
3. **Check qualification keywords** (contains any qualification keyword)
4. **Default rule**: If all caps and ≤10 characters → treat as qualification
5. **If unclear**: Context-based decision

Step 4: Multi-Value Handling
- **Join multiple values** in same category with commas
- **Set null** if no values found for a category

Complete Examples

| Original String | Split Parts | Qualification | Institute | Country |
|----------------|-------------|---------------|-----------|---------|
| `Diploma in Business Administration, LCCI Level 3, Private Secretary's Diploma, LCCI` | [Diploma in Business Administration] [LCCI Level 3] [Private Secretary's Diploma] [LCCI] | Diploma in Business Administration, LCCI Level 3, Private Secretary's Diploma, LCCI | null | null |
| `O' levels, Diploma in Business Management (distance learning), Glasgow, Dublin` | [O' levels] [Diploma in Business Management (distance learning)] [Glasgow] [Dublin] | O' levels, Diploma in Business Management (distance learning) | Glasgow, Dublin | null |
| `Diploma in Financial Informatics, Nanyang Polytechnic, CPA, SHRI Academy` | [Diploma in Financial Informatics] [Nanyang Polytechnic] [CPA] [SHRI Academy] | Diploma in Financial Informatics, CPA | Nanyang Polytechnic, SHRI Academy | null |
| `Bachelor of Arts, New Era University, Quezon City, Philippines` | [Bachelor of Arts] [New Era University] [Quezon City] [Philippines] | Bachelor of Arts | New Era University, Quezon City | Philippines |
| `Diploma in Business Administration (HRM), PSB Academy, CHRM, SHRI, Nitec in Office Skills, ITE` | [Diploma in Business Administration (HRM)] [PSB Academy] [CHRM] [SHRI] [Nitec in Office Skills] [ITE] | Diploma in Business Administration (HRM), CHRM, Nitec in Office Skills | PSB Academy, SHRI, ITE | null |



# Complete Chronological Qualification Data Parsing Logic Flow

## Master Algorithm: Comma-Based Branch System

### Initial Step: Count Commas
- Count total commas in the string
- Branch to appropriate handling logic based on comma count

---

## Branch 1: **ZERO COMMAS (0)**

### Sub-Branch 1A: Count Slashes
- If **0 slashes**: Entire string = Qualification only
 - Result: `(string, null, null)`

### Sub-Branch 1B: 1 Slash Found
- Split string by the single slash into 2 parts
- **Part 1**: Always qualification
- **Part 2**: Analyze for institute and country
 - Check if contains institution keywords (`university`, `college`, `academy`, `institute`, `school`)
 - If YES: Part 2 = Institute
 - Extract country if pattern like "University of [Country]" exists
 - Result: `(part1, part2, extracted_country_or_null)`

### Sub-Branch 1C: 2+ Slashes Found
- Split by all slashes into multiple parts
- **Part 1**: Always qualification
- **Remaining parts**: Apply same classification logic as 3+ comma case
- Use priority-based keyword matching for each part

---

## Branch 2: **ONE COMMA (1)**

### Sub-Branch 2A: No Slashes (1 comma, 0 slashes)
- Split by comma into exactly 2 parts
- **Part 1**: Always qualification
- **Part 2**: Apply classification rules in priority order:
 1. **Institute Check**: Contains `university|college|academy|institute|school|polytechnic` → Institute
 2. **Country Check**: Contains known country/city names → Country
 3. **Qualification Check**: Contains qualification keywords → Additional qualification
 4. **Default**: Treat as qualification
- Result: Distribute parts accordingly

### Sub-Branch 2B: With Slashes (1 comma, 1+ slashes)
- Split by both comma AND slashes to get 3+ parts
- **Part 1**: Always qualification
- **Parts 2+**: Apply advanced classification:
 1. **Institute Priority**: Contains institution keywords OR "ALL CAPS + offered pattern"
 2. **Country Priority**: Contains country/city names
 3. **Qualification Priority**: Contains qualification keywords
 4. **Default**: All caps ≤10 characters → Qualification
- Group classified parts by category and join with commas

---

## Branch 3: **TWO COMMAS (2)**

### Standard 3-Part Processing
- Split by commas into exactly 3 parts
- **Part 1**: Always qualification (no analysis needed)
- **Parts 2 & 3**: Apply priority-based classification:

#### Priority Classification System:
1. **Highest Priority - Country Detection**:
  - Exact match: `India`, `China`, `Philippines`, `Netherlands`, `Singapore`
  - Partial match: Contains country names within the part
  - Usually appears as last part

2. **Second Priority - Institute Detection**:
  - Institution types: `University`, `College`, `Polytechnic`, `Academy`, `Institute`, `School`
  - Known institutes: `ITE`, `PSB`, `MDIS`, `ACCA`, `SHRI`, `NCOI`
  - Specific names: `Singapore Polytechnic`, `Nanyang Polytechnic`

3. **Third Priority - Qualification Detection**:
  - Degree types: `Bachelor`, `Master`, `Diploma`, `Certificate`
  - Education levels: `O Level`, `A Level`, `N Level`, `NTC`, `Nitec`, `WSQ`
  - Professional codes: `SPM`, `LCCI`, `CPA`, `CHRM`, `CATS`

4. **Default Rule**: All caps and ≤10 characters → Treat as qualification

### Multi-Value Handling:
- If multiple parts classify as same category, join with commas
- Set null for categories with no matches

---

## Branch 4: **THREE+ COMMAS (3+)**

### Advanced Multi-Delimiter Processing

#### Step 4.1: Split by Multiple Delimiters
- Split by both commas AND slashes using regex pattern
- Clean each part: remove extra spaces, quotes, brackets
- Filter out empty parts

#### Step 4.2: Initialize Category Collections
- Create empty lists for: qualifications, institutes, countries
- Add first part to qualifications list automatically

#### Step 4.3: Advanced Classification Logic
For each remaining part, apply this hierarchical decision tree:

##### Level 1: Country Detection (Highest Priority)
- **Exact Match**: Part exactly matches known country names
- **Partial Match**: Part contains country names as substring
- **Location Indicators**: "Quezon City" (indicates Philippines)
- **Action**: Add to countries list and skip further checks

##### Level 2: Institute Detection (Second Priority)
- **Institution Types**: Contains `university`, `college`, `polytechnic`, `academy`, `institute`, `school`
- **Known Abbreviations**: Matches `ITE`, `PSB`, `MDIS`, `SHRI`, `ACCA`
- **City Names**: `Glasgow`, `Dublin`, `Rotterdam` (treat as institutes)
- **Special Pattern**: Contains "ALL CAPS + offered" pattern
- **Action**: Add to institutes list

##### Level 3: Qualification Detection (Third Priority)
- **Degree Keywords**: Contains `bachelor`, `master`, `diploma`, `certificate`, `phd`
- **Education Levels**: Contains `level`, `ntc`, `nitec`, `wsq`
- **Certifications**: Contains `certified`, `advanced`, `higher`, `foundation`
- **Professional Codes**: Matches `SPM`, `LCCI`, `CPA`, `CHRM`, `CATS`, `ACTA`
- **Action**: Add to qualifications list

##### Level 4: Default Classification Rules
- **All Caps Rule**: If part is all uppercase AND ≤10 characters → Qualification
- **Ambiguous Cases**: When multiple rules could apply, use context and position
- **Final Fallback**: Default to qualification if no clear match

#### Step 4.4: Multi-Value Assembly
- **Join Strategy**: Combine multiple values in each category with comma separation
- **Null Handling**: Set category to null if no items classified
- **Order Preservation**: Maintain relative order of items within each category

#### Step 4.5: Special Case Handling
- **Mixed Delimiters**: Handle combinations like "item1, item2/item3, item4"
- **Nested Information**: Extract relevant parts from complex strings
- **Duplicate Prevention**: Avoid adding same item to multiple categories

---

## Enhanced Keyword Database

### Qualification Keywords (Comprehensive)
- **Academic Degrees**: Bachelor, Master, Diploma, Certificate, Degree, PhD, Doctorate, Advanced Diploma
- **Education Levels**: O Level, A Level, N Level, NTC, Nitec, WSQ, Foundation
- **Professional Certifications**: Certified, Advanced, Higher, Foundation
- **Industry Codes**: SPM, WPLN, CATS, ACTA, LCCI, CPA, CHRM
- **Special Qualifications**: Private Secretary, Accounting Technicians

### Institute Keywords (Comprehensive)
- **Institution Types**: University, College, Polytechnic, Academy, Institute, School, Poly
- **Known Institutions**: ITE, PSB, MDIS, ACCA, SHRI, NCOI
- **Specific Names**: Singapore Polytechnic, Republic Poly, Nanyang Polytechnic, New Era University, PSB Academy, Tunku Abdul Rahman College
- **International**: University of Rajasthan, Anna University, Xi'an Shiyou University

### Geographic Keywords (Comprehensive)
- **Countries**: India, China, Philippines, The Netherlands, Netherlands, Singapore
- **Cities as Location Indicators**: Glasgow, Dublin, Quezon City, Rotterdam, Dumaguete City
- **Regional Indicators**: Asia, Europe (if context requires)

---

## Quality Assurance Rules

### Validation Checks
1. **Completeness**: Ensure first part always goes to qualification
2. **Consistency**: Verify classification logic applied uniformly
3. **Contextual**: Consider part position and surrounding context
4. **Flexibility**: Handle variations in formatting and terminology

### Edge Case Management
1. **Ambiguous All-Caps**: Use surrounding context to decide
2. **Multiple Categories**: If part fits multiple categories, use priority system
3. **Incomplete Information**: Handle missing delimiters gracefully
4. **Format Variations**: Account for different punctuation styles

This detailed logic flow provides a systematic approach to handle all possible comma and slash combinations while maintaining accuracy and consistency in classification.

In [24]:
# Count unique values including nulls
unique_qualifications = student_profiles['HIGHEST QUALIFICATION'].value_counts(dropna=False).reset_index()

# Rename columns for clarity
unique_qualifications.columns = ['Highest Qualification', 'Count']

# Display the result
print("Unique values in 'HIGHEST QUALIFICATION' column (including nulls):")
display(unique_qualifications)


Unique values in 'HIGHEST QUALIFICATION' column (including nulls):


Highest Qualification  Count
0                Degree    137
1           Certificate     87
2               Diploma     71
3                Master     11
4                            1

In [25]:
# Check for rows where 'HIGHEST QUALIFICATION' is null or empty
null_qualification_rows = student_profiles[student_profiles['HIGHEST QUALIFICATION'].isnull() | (student_profiles['HIGHEST QUALIFICATION'] == ' ') | (student_profiles['HIGHEST QUALIFICATION'] == '')]

print("Rows with null or empty 'HIGHEST QUALIFICATION':")
display(null_qualification_rows)

Rows with null or empty 'HIGHEST QUALIFICATION':


STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
128  2102-063/013      F        NaN   NaN         Y   

    COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
128                      Vietnam  24/07/1989                         

    NAME OF QUALIFICATION AND INSTITUTION DATE ATTAINED HIGHEST QUALIFICATION  \
128                                                                2016-06-06   

            DESIGNATION    COMMENCEMENT DATE      COMPLETION DATE  \
128  Accounts Executive  2022-04-18 00:00:00  2022-09-14 00:00:00   

    FULL-TIME OR PART-TIME COURSE FUNDING REGISTRATION FEE PAYMENT MODE  \
128              Part-Time     Individual              107         Nets   

     COURSE FEE  Comma Count  
128         888            0

Since only 1 row have got null values for  Highest Qualification, we will drop it.

#### Name of Qualification and Institution
---

In [26]:
# Split 'NAME OF QUALIFICATION AND INSTITUTION' into 'QUALIFICATION_NAME' and 'INSTITUTION_NAME'
# Use regex to split on either comma or slash
student_profiles[['QUALIFICATION_NAME', 'INSTITUTION_NAME']] = student_profiles['NAME OF QUALIFICATION AND INSTITUTION'].str.split(r',|/', n=1, expand=True)

# Display the new columns and the original column to verify
display(student_profiles[['NAME OF QUALIFICATION AND INSTITUTION', 'QUALIFICATION_NAME', 'INSTITUTION_NAME']].head())

NAME OF QUALIFICATION AND INSTITUTION  \
0                                                SPM   
1                  Certificate in Office Skills, ITE   
2  Bachelor of Business Administration, Universit...   
3  Office Management Diploma, NCOI Rotterdam, The...   
4  Diploma in Business Admininstration, LCCI Leve...   

                    QUALIFICATION_NAME  \
0                                  SPM   
1         Certificate in Office Skills   
2  Bachelor of Business Administration   
3            Office Management Diploma   
4  Diploma in Business Admininstration   

                                   INSTITUTION_NAME  
0                                              None  
1                                               ITE  
2                    University of Rajasthan, India  
3                   NCOI Rotterdam, The Netherlands  
4   LCCI Level 3, Private Secretary's Diploma, LCCI

In [27]:
# Check unique values in the 'QUALIFICATION_NAME' column
unique_qualification_names = student_profiles['QUALIFICATION_NAME'].value_counts().reset_index()
unique_qualification_names.columns = ['Qualification Name', 'Count']
print("Unique values in 'QUALIFICATION_NAME' column:")
display(unique_qualification_names) # Displaying head as there might be many unique values

# Check unique values in the 'INSTITUTION_NAME' column
unique_institution_names = student_profiles['INSTITUTION_NAME'].value_counts().reset_index()
unique_institution_names.columns = ['Institution Name', 'Count']
print("\nUnique values in 'INSTITUTION_NAME' column:")
display(unique_institution_names) # Displaying head as there might be many unique values

Unique values in 'QUALIFICATION_NAME' column:


Qualification Name  Count
0                                                  SPM     15
1                                             O' level     15
2                   Diploma in Business Administration      7
3                  Bachelor of Business Administration      6
4                                            O' levels      6
..                                                 ...    ...
171                   Bachelor of Business (Marketing)      1
172  Bachelor of Commerce in Management and Marketi...      1
173                       Bachelor of Arts (Sociology)      1
174                      Bachelor of Arts with Honours      1
175                         Degree of Computer Science      1

[176 rows x 2 columns]


Unique values in 'INSTITUTION_NAME' column:


Institution Name  Count
0                                      ITE     13
1       \nNational University of Singapore     12
2                     \nMurdoch University      8
3                              PSB Academy      6
4                    Singapore Polytechnic      6
..                                     ...    ...
129                  \nUniversity of Delhi      1
130                 \nUniversity of Malaya      1
131    \nNational Technological University      1
132                  \nNanyang Polytechnic      1
133  \nQueensland University of Technology      1

[134 rows x 2 columns]

#### qualification

In [28]:
import re

# Step 1: Clean and lowercase the qualification names
def clean_qualification(name):
    if not isinstance(name, str):
        return ''
    # Normalize quotes and strip
    name = name.lower().strip()
    name = re.sub(r"[‘’'`\"]", '', name)
    # Standardize O Level variants
    if re.match(r"^(o\s*level|o's?\s*level|o\s*levels)$", name):
        return 'o level'
    # Standardize A Level variants
    if re.match(r"^(a\s*level|a's?\s*level|a\s*levels)$", name):
        return 'a level'
    # Standardize N Level variants
    if re.match(r"^(n\s*level|n's?\s*level|n\s*levels)$", name):
        return 'n level'
    # Treat empty or dash as blank
    if name in ['', '-']:
        return ''
    return name

student_profiles['Qualification Clean'] = student_profiles['QUALIFICATION_NAME'].apply(clean_qualification)

# Step 2: Define field keywords and categories (split school levels)
field_keywords = {
    'Business': ['business', 'commerce', 'marketing', 'accounting', 'finance', 'management', 'entrepreneurship', 'accountancy'],
    'IT': ['information technology', 'infocomm', 'computing', 'computer science', 'ict', 'software', 'cybersecurity', 'data', 'ai', 'machine learning'],
    'Engineering': ['engineering', 'mechanical', 'electrical', 'civil', 'aerospace', 'electronics', 'mechatronics'],
    'Healthcare': ['nursing', 'pharmacy', 'biomedical', 'healthcare', 'medical', 'life sciences'],
    'Design & Media': ['design', 'media', 'animation', 'visual', 'graphic', 'interior', 'fashion'],
    'Education': ['education', 'teaching', 'early childhood'],
    'Hospitality': ['hospitality', 'tourism', 'culinary', 'hotel'],
    'Sciences': ['science', 'chemistry', 'biology', 'physics', 'mathematics'],
    'Social Sciences': ['psychology', 'sociology', 'social work', 'counselling'],
    'Law': ['law', 'legal'],
    'Arts': ['art', 'communication', 'public relations', 'journalism', 'media', 'performing arts'],
    'O Level': ['o level'],
    'A Level': ['a level'],
    'N Level': ['n level'],
    'Others': []
}

# Step 3: Categorize based on keywords
def categorize_field(qualification):
    if qualification == '':
        return 'Others'
    for category, keywords in field_keywords.items():
        if any(keyword in qualification for keyword in keywords):
            return category
    return 'Others'

# Step 4: Apply categorization
student_profiles['Qualification Field'] = student_profiles['Qualification Clean'].apply(categorize_field)

# Step 5: Group and display counts
field_counts = student_profiles['Qualification Field'].value_counts().reset_index()
field_counts.columns = ['Qualification Field', 'Count']

print("Qualification Field Categories Count:")
display(field_counts)




Qualification Field Categories Count:


Qualification Field  Count
0             Business    140
1               Others     48
2              O Level     24
3             Sciences     20
4                 Arts     14
5              N Level     14
6          Engineering     11
7                   IT      8
8          Hospitality      8
9            Education      7
10      Design & Media      5
11             A Level      3
12     Social Sciences      3
13          Healthcare      2

As we can see, business is the most commonly found field in this dataset. Do note that data are categorised based on only field, meaning in business category, it may include, degree, diploma or even master.

In [29]:
# Filter and display qualifications categorized as 'Others'
others_df = student_profiles[student_profiles['Qualification Field'] == 'Others']

# See unique "qualification name" entries in 'Others'
unique_others = others_df['QUALIFICATION_NAME'].value_counts().reset_index()
unique_others.columns = ['Qualification Name', 'Count']

print("Qualifications Categorized as 'Others':")
display(unique_others)


Qualifications Categorized as 'Others':


Qualification Name  Count
0                                                SPM     15
1                       Certificate in Office Skills      5
2                                                BBA      3
3               Graduate Diploma in Systems Analysis      3
4          Degree of Bachelor of Development Studies      2
5          WSQ Higher Certificate in Human Resources      2
6                                              NTC 2      2
7              Certificate in Payroll Administration      1
8                   Nitec in Service Skills (Office)      1
9                  Bachelor of Public Administration      1
10                  Diploma in Financial Informatics      1
11                                                        1
12                                                 -      1
13                       Diploma in Maritime Studies      1
14                          Basic Certificate in HRM      1
15                Certificate in Opticianry Practice      1
16                                Certificate in HRM      1
17                             Bachelor of Economics      1
18                                             FCHRM      1
19                           Graduate Diploma in TRM      1
20                         Specialist Diploma in HRM      1
21                 NTC-2 Chemical Process Technology      1
22  Degree of Master of English as a Second Language      1

Can have SPM as a separate category.

#### Institution

In [30]:
# Updated function to categorize institution name
def categorize_institution(name):
    if not isinstance(name, str):
        return 'Other'

    # Clean and standardize
    clean_name = name.strip().replace('\n', '').upper()

    # List of known Singapore universities (you can expand this list as needed)
    singapore_universities = [
        'NUS', 'NATIONAL UNIVERSITY OF SINGAPORE',
        'NTU', 'NANYANG TECHNOLOGICAL UNIVERSITY',
        'SUSS', 'SINGAPORE UNIVERSITY OF SOCIAL SCIENCES',
        'SIT', 'SINGAPORE INSTITUTE OF TECHNOLOGY',
        'SMU', 'SINGAPORE MANAGEMENT UNIVERSITY'
    ]

    # Check for known universities
    if any(uni in clean_name for uni in singapore_universities) or 'UNIVERSITY' in clean_name:
        return 'University'
    elif 'POLYTECHNIC' in clean_name:
        return 'Polytechnic'
    elif clean_name == 'ITE':
        return 'ITE'
    else:
        return 'Other'

# Apply cleaning and categorization
student_profiles['Institution Clean'] = student_profiles['INSTITUTION_NAME'].astype(str).str.strip().str.replace('\n', '')
student_profiles['Institution Clean'] = student_profiles['Institution Clean'].str.upper()

student_profiles['Institution Category'] = student_profiles['Institution Clean'].apply(categorize_institution)

# Count by category
institution_counts = student_profiles['Institution Category'].value_counts().reset_index()
institution_counts.columns = ['Institution Category', 'Count']

print("Counts by Institution Category:")
display(institution_counts)


Counts by Institution Category:


Institution Category  Count
0                Other    137
1           University    130
2          Polytechnic     27
3                  ITE     13

In [31]:
# Filter rows where the Institution Category is 'Other'
other_institutions = student_profiles[student_profiles['Institution Category'] == 'Other']

print("Institutions categorized as 'Other':")
display(other_institutions[['INSTITUTION_NAME', 'Institution Clean']].drop_duplicates())


Institutions categorized as 'Other':


INSTITUTION_NAME  \
0                                                 None   
3                      NCOI Rotterdam, The Netherlands   
4      LCCI Level 3, Private Secretary's Diploma, LCCI   
5                                             Malaysia   
17                                         PSB Academy   
19              Hospitality and Event Management, MDIS   
21    Diploma in Business Management (distance lear...   
22                                           ACCA, SPM   
23                                    ACCA, 'O' levels   
24                                       N Level, WPLN   
25                                           'N' Level   
26                                                MDIS   
28                                                 NYP   
32    WPLN, Fundamentals Certificate in Early Child...   
35                                       SIPMM Academy   
36                                   SHRI Academy, SPM   
43                                              Kaplan   
65                                              SHATEC   
68                                                 UOL   
76                             Ascend Education Centre   
83                                                HCS    
85    Certificate in Business Studies (Business Adm...   
86    LCCI Level 2 Certificate in Book-keeping and ...   
88    Foundation Certificate in Human Resource Mana...   
93         Foundation Certificate in HRM, SHRI Academy   
97                                                LCCI   
99                                         HLI Academy   
102          Computer Networking & Communications, ITE   
103                                               SHRI   
118                                          PSB, ACTA   
119                          Lyceum of the Philippines   
122   Institute of Maritime & Business Management (...   
129                                       PSB Academy    
134   PSB Academy / CHRM, SHRI / Nitec in Office Sk...   
144     St Paul College of Dumaguete City, Philippines   
151                  Kaplan Higher Education Institute   
163                       National Productivity Board    
169                                       SHRI Academy   
177                                                MIS   
179                                Republic Poly, WPLN   
189                   Marketing Institute of Singapore   
193                                                SIM   
194                                                MMU   
195                Private Secretary Certificate, LCCI   
197                                   Achiever Academy   
207                                    SSG / 'A' level   
210                      The Logistics Academy Pte Ltd   
226                   \nSystematic College Johor Bahru   
227                    \nYangon Institute of Economics   
229                \nKaplan Higher Education Institute   
231  \nInternational Business and Management Educat...   
233                             \nSHRI Academy Pte Ltd   
237          \nRoyal Melbourne Institute of Technology   
238              \nSingapore Human Resource Management   

                                     Institution Clean  
0                                                 NONE  
3                      NCOI ROTTERDAM, THE NETHERLANDS  
4      LCCI LEVEL 3, PRIVATE SECRETARY'S DIPLOMA, LCCI  
5                                             MALAYSIA  
17                                         PSB ACADEMY  
19              HOSPITALITY AND EVENT MANAGEMENT, MDIS  
21   DIPLOMA IN BUSINESS MANAGEMENT (DISTANCE LEARN...  
22                                           ACCA, SPM  
23                                    ACCA, 'O' LEVELS  
24                                       N LEVEL, WPLN  
25                                           'N' LEVEL  
26                                                MDIS  
28                                                 NYP  
32   WPLN, FUNDAMENTALS CERTIFICATE IN EARLY CHIL

In [32]:
# Filter rows where Institution Category is 'Other'
other_institutions = student_profiles[student_profiles['Institution Category'] == 'Other']

# Get unique cleaned institution names in the 'Other' category
unique_other_institutions = other_institutions['Institution Clean'].unique()

print("All institution names categorized as 'Other':")
for inst in unique_other_institutions:
    print(inst)


All institution names categorized as 'Other':
NONE
NCOI ROTTERDAM, THE NETHERLANDS
LCCI LEVEL 3, PRIVATE SECRETARY'S DIPLOMA, LCCI
MALAYSIA
PSB ACADEMY
HOSPITALITY AND EVENT MANAGEMENT, MDIS
DIPLOMA IN BUSINESS MANAGEMENT (DISTANCE LEARNING), GLASGOW, DUBLIN
ACCA, SPM
ACCA, 'O' LEVELS
N LEVEL, WPLN
'N' LEVEL
MDIS
NYP
WPLN, FUNDAMENTALS CERTIFICATE IN EARLY CHILDHOOD CARE AND EDUCATION
SIPMM ACADEMY
SHRI ACADEMY, SPM
KAPLAN
SHATEC
UOL
ASCEND EDUCATION CENTRE
HCS
CERTIFICATE IN BUSINESS STUDIES (BUSINESS ADMINISTRATION) LEVEL IV, TUNKU ABDUL RAHMAN COLLEGE
LCCI LEVEL 2 CERTIFICATE IN BOOK-KEEPING AND ACCOUNTS / LCCI LEVEL 3 BUSINESS STATISTICS
FOUNDATION CERTIFICATE IN HUMAN RESOURCE MANAGEMENT, SHRI ACADEMY
FOUNDATION CERTIFICATE IN HRM, SHRI ACADEMY
LCCI
HLI ACADEMY
COMPUTER NETWORKING & COMMUNICATIONS, ITE
SHRI
PSB, ACTA
LYCEUM OF THE PHILIPPINES
INSTITUTE OF MARITIME & BUSINESS MANAGEMENT (IMBM)
PSB ACADEMY / CHRM, SHRI / NITEC IN OFFICE SKILLS, ITE
ST PAUL COLLEGE OF DUMAGUETE CITY,

## Data Cleaning
---

### Data Cleaning Steps for Student Profile

#### 1. Student ID Decomposition

* Split `STUDENT_ID` into three new columns:

  * `COURSE_ID`
  * `INTAKE_NO`
  * `INDEX_NO`
* Format: `<COURSE_ID>-<INTAKE_NO>/<INDEX_NO>`

#### 2. Remove Unmatched Semester Results

* Remove entries from `semester_results` if the student does not exist in the `student profile` table.

#### 3. Gender

* Clean - Have 2 unique values ('M' and 'F')

#### 4. Residential Status

* Combine `SG CITIZEN`, `SG PR`, and `FOREIGNER` into a single column: `RESIDENTIAL_STATUS`.

  * Logic: Only one of the three columns contains "Y"; others are null.
  * New values: `Singapore Citizen`, `PR`, or `Foreigner`.

#### 5. Nationality

* Rename `COUNTRY OF OTHER NATIONALITY` to `NATIONALITY`.
* Replace null or blank values with: `Singapore`.
* Replace Malaysian with Malaysia

#### 6. Date of Birth (`DOB`)

* Not in datetime format
* Some rows are in short abbreviation for month instead of numbers

#### 7. Highest Qualification

* Drop rows where value is `" "`

#### 8. Name of Qualification and Institution

* Split `NAME OF QUALIFICATION AND INSTITUTION` into two parts:

  * `QUALIFICATION_NAME` (text before the first comma)
  * `INSTITUTION_NAME` (text after the first comma)
  * Standardise qualification Name
  * Standardise institution name

#### 9. Date Attained Highest Qualification

* Check if the date is realistic compared to `DOB`.

#### 10. Designation

* Leave as-is for now.

#### 11. Commence Date & Completion Date

* If `COMPLETION DATE` is null, it may indicate the student has not completed the course.

  * Consider keeping the null value but add a column like `COURSE_COMPLETED` (True/False) for clarity.

#### 12. Full-time or Part-time

* Standardize values:

  * Check for null values.
  * Correct spelling errors (e.g., `fulltime` → `Full-time`).

#### 13. Course Funding Type

* Standardize values to one of the following:

  * `Individual`
  * `Individual - SFC (SkillsFuture Credit)`
  * `Sponsored`
  * `Individual - Waived App Fee`
* Check for spelling errors and inconsistencies.

#### 14. Registration Fee

* Remove `$` symbol.
* Convert to float with 2 decimal places.

#### 15. Payment Mode

* Standardize text and check for spelling mistakes.

#### 16. Course Fee

* Remove `$` symbol.
* Convert to float with 2 decimal places.

---


## Exporting Data
---